### PR TITLE
Add the ability to create one-to-one 'join_on_key'-type links to the GUI link editor

### PR DESCRIPTION
**Pull request by jfoster17**
## Description

This is a combination of a plug-in and core changes to the link_editor to enable the creation, display, and removal of join_on_key type links (managed through a `Join_Link` object). This address the immediate need in #2215.

### CODE REVIEW

**Comment by jfoster17** [comment]
Hey @astrofrog -- can you take a look at this solution? It's not... ideal, but it basically works. Basically I have a `Join_Link(LinkCollection)` object that represents the join_on_key operation and then the link_manager knows how to use this object to add or remove the appropriate `_key_joins`. Here are some of my concerns.

1. Because the key joins are not the same thing at the Join_Link object, if users (or code) directly calls join_on_key, this won't create a Join_Link and so there will be no representation of the join in the UI.
2. Because Join_Links are funny kinds of things, I end up with `data_collection._link_manager_external_links` holding `Join_Link`s, but `Join_Link`s do NOT show up in `data_collection.links`
3. Currently I have this limited to "Joining on single components". (I don't see a fundamental reason I couldn't extend it to multiple components or one-to-many joins I just haven't done that yet).
4. The display in the link manager is a little funny -- I draw these Join_Links as dashed lines, but a new Join_Link does not show up as dashed right away, only after closes and re-opening the link manager. This can probably be fixed in `link_editor.qt.data_graph` but I haven't dug into it a lot (it's not obvious what to display if two datasets are both joined_on_key and glued through a traditional link for some reason). 
5. I set up Join_Link as a plugin originally, hoping that I might be able to do things all as a plug-in. Now I have core components assume this plugin exists, but is obviously not ideal (a user might disable the plug-in). Presumably I should move it to core.link_helpers?

File: glue/core/link_manager.py [hunk 1]
```diff
@@ -30,6 +30,7 @@
 from glue.core.subset import Subset
 from glue.utils import unbroadcast
 from glue.core.coordinate_helpers import dependent_axes
+from glue.plugins.join_on_key.link_helpers import Join_Link
```
Comment (astrofrog) [inline glue/core/link_manager.py:33]: I think the class should probably just be called JoinLink (no underscore)

File: glue/dialogs/link_editor/qt/data_graph.py [hunk 2]
```diff
@@ -162,8 +166,12 @@ def get_connections(dc_links):
     for link in dc_links:
         data1 = link.data1
         data2 = link.data2
+        if link.join_link:
+            join_link = True
+        else:
+            join_link = False
         if (data1, data2) not in links and (data2, data1) not in links:
-            links.append((data1, data2))
+            links.append((data1, data2, join_link))
```
Comment (astrofrog) [inline glue/dialogs/link_editor/qt/data_graph.py:174]: I wonder if instead of having join_link be a boolean we should have the third argument be a link_type which could be a string, then it could be currently either 'join' or some name we pick for the current kind of linking (I don't have strong opinions on what that should be currently, could be e.g. 'value' or something like that.

File: glue/plugins/join_on_key/link_helpers.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+from glue.config import link_helper
+from glue.core.link_helpers import LinkCollection
+
+
+@link_helper(category="Join")
+class Join_Link(LinkCollection):
+    cid_independent = False
+
+    display = "Join on ID"
+    description = "Join two datasets on a common indetifier or index. \
```
Comment (astrofrog) [inline glue/plugins/join_on_key/link_helpers.py:10]: typo in 'indetifier'

File: glue/plugins/join_on_key/link_helpers.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+This is similar to a database join in that subsets defined \
+on one dataset can always propogate through joins."
```
Comment (astrofrog) [inline glue/plugins/join_on_key/link_helpers.py:12]: I would perhaps give a little more detail and an example - to show how it differs to a classic link?

**Review by astrofrog — CHANGES_REQUESTED** [review]
This is looking good so far! A few comments below.

> Because the key joins are not the same thing at the Join_Link object, if users (or code) directly calls join_on_key, this won't create a Join_Link and so there will be no representation of the join in the UI.

To some extent, join_on_key should perhaps become deprecated API once we are confident it can be treated as a regular link - that is, in future maybe creating a ``JoinLink`` should be the preferred way to do it. So I wouldn't worry about this.

> Because Join_Links are funny kinds of things, I end up with data_collection._link_manager_external_links holding Join_Links, but Join_Links do NOT show up in data_collection.links

Ok I don't know immediately why that would be the case, but we should be able to get that to work in future.

> Currently I have this limited to "Joining on single components". (I don't see a fundamental reason I couldn't extend it to multiple components or one-to-many joins I just haven't done that yet).

Sounds good

> The display in the link manager is a little funny -- I draw these Join_Links as dashed lines, but a new Join_Link does not show up as dashed right away, only after closes and re-opening the link manager. This can probably be fixed in link_editor.qt.data_graph but I haven't dug into it a lot (it's not obvious what to display if two datasets are both joined_on_key and glued through a traditional link for some reason).

Ok yes we can look into this later

> I set up Join_Link as a plugin originally, hoping that I might be able to do things all as a plug-in. Now I have core components assume this plugin exists, but is obviously not ideal (a user might disable the plug-in). Presumably I should move it to core.link_helpers?

Yes I think maybe it should just be a core link type.

The bottom line is that I'm happy for this to go into the core package, and I think it was overdue that joins become normal links - thanks @jfoster17!

File: glue/core/link_manager.py (hunk 1 above)
Comment (jfoster17) [reply glue/core/link_manager.py]: Good point, I have renamed.

File: glue/dialogs/link_editor/qt/data_graph.py (hunk 2 above)
Comment (jfoster17) [reply glue/dialogs/link_editor/qt/data_graph.py]: That makes sense. For now I have 'join' and 'value' as the two link_types

File: glue/plugins/join_on_key/link_helpers.py (hunk 3 above)
Comment (jfoster17) [reply glue/plugins/join_on_key/link_helpers.py]: Fixed

File: glue/plugins/join_on_key/link_helpers.py (hunk 4 above)
Comment (jfoster17) [reply glue/plugins/join_on_key/link_helpers.py]: Taking inspiration from [the docs](http://docs.glueviz.org/en/stable/gui_guide/merging.html) I now describe this as:

> Join two datasets on a common ID. Other links 
> in glue connect data columns (two datasets have 'age' columns but 
> the rows are different objects), while Join on ID connects the same 
> rows/items across two datasets.

There isn't a great deal of space in the UI for a longer description, and concise is better anyway. Do you think this is sufficient?

**Comment by jfoster17** [comment]
I assume you would like to be rebase and add a changelog entry when this is good-to-go?

**Comment by astrofrog** [comment]
Yes that would be great, thanks!

File: glue/plugins/join_on_key/link_helpers.py (hunk 4 above)
Comment (jfoster17) [reply glue/plugins/join_on_key/link_helpers.py]: @astrofrog -- if you approve of this new description I think this is good-to-go?
